core/src/subgraph/provider_with_names.rs

```diff
@@ -9,45 +9,109 @@ pub struct SubgraphProviderWithNames<P, S> {
     logger: slog::Logger,
     provider: Arc<P>,
     store: Arc<S>,
-}
-
-impl<P, S> Clone for SubgraphProviderWithNames<P, S> {
-    fn clone(&self) -> Self {
-        Self {
-            logger: self.logger.clone(),
-            provider: self.provider.clone(),
-            store: self.store.clone(),
-        }
-    }
+    node_id: NodeId,
+    _deployment_event_stream_cancel_guard: CancelGuard, // cancels on drop
 }
 
 impl<P, S> SubgraphProviderWithNames<P, S>
 where
     P: SubgraphProviderTrait,
-    S: Store,
+    S: SubgraphDeploymentStore,
 {
-    pub fn new(logger: slog::Logger, provider: Arc<P>, store: Arc<S>) -> Self {
```

Comment:
It doesn't revert because `main` is not using `wait`. I did like the style of separating the calls though, https://github.com/graphprotocol/graph-node/pull/606#discussion_r237028833 is about this same thing.